src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -0,0 +1,143 @@
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMSource;
+import hudson.Extension;
+import hudson.model.Item;
+import hudson.plugins.git.BranchSpec;
+import hudson.scm.SCM;
+import hudson.scm.SCMDescriptor;
+import hudson.util.FormValidation.Kind;
+import jenkins.plugins.git.GitBranchSCMHead;
+import jenkins.scm.api.*;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+import static org.eclipse.jgit.lib.Constants.*;
+
+public class BitbucketSCMFileSystem extends SCMFileSystem {
+
+    private final BitbucketFilePathClient client;
+    private final String ref;
+
+    protected BitbucketSCMFileSystem(BitbucketFilePathClient client, @Nullable SCMRevision scmRevision,
+                                     @Nullable String ref) {
+        super(scmRevision);
+        this.client = client;
+        this.ref = ref;
+    }
+
+    @Override
+    public SCMFile getRoot() {
+        return new BitbucketSCMFile(client, ref);
+    }
+
+    // We do not provide this information in the REST response, so this is undefined.
+    @Override
+    public long lastModified() throws IOException, InterruptedException {
+        return 0;
+    }
+
+    @Extension
+    public static class BuilderImpl extends SCMFileSystem.Builder {
+
+        @Inject
+        BitbucketClientFactoryProvider clientFactoryProvider;
+        @Inject
+        JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
+        @Inject
+        BitbucketPluginConfiguration pluginConfiguration;
+
+        @Override
+        public SCMFileSystem build(Item item, SCM scm,
+                                   @CheckForNull SCMRevision scmRevision) throws IOException, InterruptedException {
+            if (!(scm instanceof BitbucketSCM)) {
+                return null;
+            }
+
+            BitbucketSCM bitbucketSCM = (BitbucketSCM) scm;
+            Optional<BitbucketServerConfiguration> maybeServerConfiguration =
+                    pluginConfiguration.getServerById(bitbucketSCM.getServerId());
+            if (!maybeServerConfiguration.isPresent() || maybeServerConfiguration.get().validate().kind == Kind.ERROR) {
+                return null;
+            }
+
+            BitbucketSCMRepository repository = bitbucketSCM.getBitbucketSCMRepository();
+
+            BitbucketFilePathClient filePathClient =
+                    clientFactoryProvider.getClient(maybeServerConfiguration.get().getBaseUrl(),
+                            jenkinsToBitbucketCredentials.toBitbucketCredentials(repository.getCredentialsId()))
+                            .getProjectClient(repository.getProjectKey())
+                            .getRepositoryClient(repository.getRepositorySlug())
+                            .getFilePathClient();
+
+            return new BitbucketSCMFileSystem(filePathClient, null, bitbucketSCM.getBranches().get(0).toString());
```

Comment:
`getBranches` can return an empty list so i don't think its ok to just do `getBranches().get(0)`

Reply:
We have already determined getBranches contains exactly one element in the `supports` method. I can add the check again to this method, but the consumer should _not_ be making changes to the SCM between calling `supports` and `build`

Reply:
Is `supports` always going to be called before `build`?

Reply:
From `SCMFileSystem.java`:

`public static SCMFileSystem of(@NonNull Item owner, @NonNull SCM scm, @CheckForNull SCMRevision rev)
            throws IOException, InterruptedException {
        Objects.requireNonNull(scm);
        SCMFileSystem fallBack = null;
        Throwable failure = null;
        for (Builder b : ExtensionList.lookup(Builder.class)) {
            if (b.supports(scm)) {
                try {
                    SCMFileSystem inspector = b.build(owner, scm, rev);`

The only way this could be subverted is if a consumer ignores the SCMFileSystem and performs the extension list lookup themselves, and then chooses to ignore the `supports` method.